misc/mdns/src/lib.rs

```diff
@@ -0,0 +1,561 @@
+                                ));
+                            } else {
+                                // Note that ideally we would use a loop instead. However as of the
+                                // writing of this code non-lexical lifetimes haven't been merged
+                                // yet, and I can't manage to write this code without having borrow
+                                // issues.
+                                task::current().notify();
+                                return Async::NotReady;
+                            }
+                        } else {
+                            return Async::Ready(MdnsPacket::Response(MdnsResponse {
+                                packet,
+                                from,
+                            }));
+                        }
+                    }
+                    Err(_) => {
+                        // Ignore errors while parsing the packet. We need to poll again for the
+                        // next packet.
+                        // Note that ideally we would use a loop instead. However as of the writing
+                        // of this code non-lexical lifetimes haven't been merged yet, and I can't
+                        // manage to write this code without having borrow issues.
+                        task::current().notify();
+                        return Async::NotReady;
+                    }
+                }
+            }
+            Ok(Async::NotReady) => (),
+            Err(_) => {
+                // Error are non-fatal and can happen if we get disconnected from example.
+                // The query interval will wake up the task at some point so that we can try again.
+            }
+        };
+
+        Async::NotReady
+    }
+}
+
+/// A valid mDNS packet received by the service.
+#[derive(Debug)]
+pub enum MdnsPacket<'a> {
+    /// A query made by a remote.
+    Query(MdnsQuery<'a>),
+    /// A response sent by a remote in response to one of our queries.
+    Response(MdnsResponse<'a>),
+    /// A request for service discovery.
+    ServiceDiscovery(MdnsServiceDiscovery<'a>),
+}
+
+/// A received mDNS query.
+pub struct MdnsQuery<'a> {
+    /// Sender of the address.
+    from: SocketAddr,
+    /// Id of the received DNS query. We need to pass this ID back in the results.
+    query_id: u16,
+    /// Queue of pending buffers.
+    send_buffers: &'a mut Vec<Vec<u8>>,
+}
+
+impl<'a> MdnsQuery<'a> {
+    /// Respond to the query.
+    ///
+    /// Pass the ID of the local peer, and the list o addresses we're listening on.
+    ///
+    /// If there are more than 2^16-1 addresses, ignores the others.
+    ///
+    /// > **Note**: Keep in mind that we will also receive this response in an `MdnsResponse`.
+    #[inline]
+    pub fn respond<TAddresses>(
+        self,
+        peer_id: PeerId,
+        addresses: TAddresses,
+        ttl: Duration,
+    ) -> Result<(), MdnsResponseError>
```

Comment:
Should responders not probe and announce on startup (cf. RFC 6762, section 8)?

Reply:
It would be really great to avoid having to figure out the IP address of our network interfaces, as there's no easy cross-platform way to do that.
If the specs allow avoiding that, it would be great.